### PR TITLE
[PS5][Driver] Supply libraries and CRT objects to the linker

### PR DESCRIPTION
**Pull request by playstation-edd**
Until now, these have been hardcoded as a downstream patches in lld. Add them to the driver so that the private patches can be removed.

PS5 only. On PS4, the equivalent hardcoded configuration will remain in the proprietary linker.

SIE tracker: TOOLCHAIN-16704

### CODE REVIEW

**Review by jmorse — COMMENTED** [review]
Just to check my understanding, things like CHECK-MAIN-CRT-SAME are going to shift the match-position that FileCheck is looking at forwards, so that we have to match in the correct order, yes?

I get the feeling that given how much of the construct-a-job logic filters for `!Relocatable` it might be worth having that as an early exit. On the other hand, the logic here is already a complicated matrix of expectations, no need to make it more complicated in the name of lines-of-code.

For the CHECK-NO... flavour of checklines, would they be better as `--check-implicit-not` arguments to FileCheck, or is it alright to rely on them being in a rigid order? (No strong opinion).

File: clang/test/Driver/ps5-linker.c [hunk 1]
```diff
@@ -97,6 +97,60 @@
...
+// RUN: %clang --target=x86_64-sie-ps5 %s -nolibc -static -### 2>&1 | FileCheck --check-prefixes=CHECK-LD,CHECK-STATIC-CRT,CHECK-STATIC-LIBCPP,CHECK-NO-LIBC,CHECK-STATIC-CORE-LIBS %s
+
+// RUN: %clang --target=x86_64-sie-ps5 %s -nostdlib++ -### 2>&1 | FileCheck --check-prefixes=CHECK-LD,CHECK-MAIN-CRT,CHECK-NO-LIBCPP,CHECK-DYNAMIC-CORE-LIBS %s
+// RUN: %clang --target=x86_64-sie-ps5 %s -nostdlib++ -shared -### 2>&1 | FileCheck --check-prefixes=CHECK-LD,CHECK-SHARED-CRT,CHECK-NO-LIBCPP,CHECK-DYNAMIC-CORE-LIBS %s
+// RUN: %clang --target=x86_64-sie-ps5 %s -nostdlib++ -static -### 2>&1 | FileCheck --check-prefixes=CHECK-LD,CHECK-STATIC-CRT,CHECK-NO-LIBCPP,CHECK-STATIC-LIBC,CHECK-STATIC-CORE-LIBS %s
+
+// CHECK-LD: {{ld(\.exe)?}}"
+// CHECK-MAIN-CRT-SAME: "crt1.o" "crti.o" "crtbegin.o"
+// CHECK-SHARED-CRT-SAME: "crti.o" "crtbeginS.o"
+// CHECK-STATIC-CRT-SAMW: "crt1.o" "crti.o" "crtbeginT.o"
```
[the remark this answers is not in the document]
Comment (playstation-edd) [reply clang/test/Driver/ps5-linker.c]: Thank you - fixed.

**Comment by playstation-edd** [comment]
> Just to check my understanding, things like CHECK-MAIN-CRT-SAME are going to shift the match-position that FileCheck is looking at forwards, so that we have to match in the correct order, yes?

Yep!

> I get the feeling that given how much of the construct-a-job logic filters for `!Relocatable` it might be worth having that as an early exit. On the other hand, the logic here is already a complicated matrix of expectations, no need to make it more complicated in the name of lines-of-code.

There's are bits of `Relocatable`-agnostic stuff in the middle (`AddLinkerInputs()`) and at the end, so I don't know how much tidier or smaller it can be made. I'm don't find it's current state especially offensive, but I could have a go at moving the new stuff to some helper functions, perhaps?

> For the CHECK-NO... flavour of checklines, would they be better as `--check-implicit-not` arguments to FileCheck, or is it alright to rely on them being in a rigid order? (No strong opinion).

Thanks - `--check-implicit-not` would indeed be better from a strictness point of view. So I have pushed this change for your consideration. After doing it, I'm not so sure it's overall better, as the regular expressions are now repeated in a number of places and I have had to carefully quote them so that lit's command parser doesn't e.g. interpret `*` as glob, and instead leaves it for `FileCheck` to interpret as a part of a regex. I have a very slight preference to revert this particular commit as I feel there's slightly more chance for error now, though I did validate with some experiments. Let me know what you think.

**Comment by pogo59** [comment]
FWIW, IMO the implicit-check-not idea was worth trying, but not worth keeping. Up to @jmorse though.

**Review by jmorse — APPROVED** [review]
> I don't find its current state especially offensive, but I could have a go at moving the new stuff to some helper functions, perhaps?

If it's manageable then it's manageable, 

> I have a very slight preference to revert this particular commit as I feel there's slightly more chance for error now, though I did validate with some experiments. Let me know what you think.

Yeah, while there might be slightly more coverage it comes at the cost of considerable uglyness, so backing that out sounds is preferable to me too.

Does the flipped order of library-option-names in 100c3c271 have an actual effect on behaviour? Just curious.

LGTM

**Comment by playstation-edd** [comment]
> Does the flipped order of library-option-names in [100c3c2](https://github.com/llvm/llvm-project/commit/100c3c271437b3b6181a5a57381363f4b6e9e841) have an actual effect on behaviour? Just curious.

No effect. I accidentally flipped the order it in the initial commit - presumably the result of iterating and experimenting along the way. The follow-up commit was to flip it back. Sorry for the distraction.

**Pull request merged**